#14 [auction] 입찰하기 기능 구현

Kminss opened the pull request:
- 5% 이상 상회입찰가면 DB저장 후 Redis에  입찰가 갱신 => 추후 DB 저장도 비동기로 처리하면 더 좋을 것 같음

- Redisson 자료구조에 값을 저장할 때, 클래스 타입 정보를 같이 저장함.
 값만 저장하기 위해서는 codec에 타입 정보클래스를 초기화 하고 해당 코덱을
 세팅해줘야함
 입찰관련 redisson repository class 를 만들어 사용하기로 변경
 추후 리팩토링 여지가 많음

 - 요청한 입찰가에 대해 포인트 검증 및 차감 추가
 
 - ToDo: 
    - 일정 여유되면 알림기능도 연동
    - redisson repository 리팩토링
    - 입찰가 포인트 차감 리팩토링
   
## 개요 📝
<!---- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->

<!---- Resolves: #(Isuue Number) -->
Resolves: #14 

## PR 유형 ✅
어떤 변경 사항이 있나요?

- [X] 새로운 기능 추가
- [ ] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## PR Checklist ✅
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [X] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

<br>

## 작업사항 🛠️
<br>

## 변경로직 ⚠️ 
<br>

### 변경전
<br>

### 변경후
<br>

## 기타

Kminss committed:
#14 [update-auction] 경매 도메인 - 입찰하기 기능 구현

 - 5% 이상 상회입찰가면 DB저장 후 Redis에  입찰가 갱신
   => 추후 DB 저장도 비동기로 처리하면 더 좋을 것 같음

 - ToDo: 일정 여윺되면 알림기능도 연동

Kminss committed:
#14 [refactor-auction] 경매 도메인 - 입찰 서비스 테스트 리팩토링

 - 사용 안하는 코드 삭제

Kminss committed:
#14 [update-auction] 경매 도메인 - 정상 응답 코드값 추가

kimhyeyun reviewed (approved):
좋슴돠~ 고생하셨습니다~~

OuOHoon commented on src/main/java/com/sparta/ticketauction/domain/auction/service/BidServiceImpl.java:
```diff
@@ -1,4 +1,61 @@
+		long increaseBidPrice = (long)(currentBidPrice * BID_PRICE_INCREASE_PERCENT);
+		long bidPrice = bidRequest.getPrice();
+
+		if (increaseBidPrice > bidPrice) {
```
입찰시 현재 유저가 포인트를 입찰 가능한만큼 보유하고 있는지 확인하는 과정이 필요하다고 생각하는데 어떻게 생각하시나요~? 아니면 다른 서비스에서 처리하실 예정일까요~?

Kminss replied:
수정했습니다. 
우선 직접 userRepo에서 point 가져와 검증 후 업데이트 하는식으로 구현했습니다.
다만 해당서비스에서 바로 하는것보다 
Point 관련된 서비스레이어에서 `포인트 결제` 라는 서비스를 구현하고 해당 서비스 클래스를 의존하는 식으로
변경해야 할 것 같습니다.

OuOHoon reviewed (changes requested):
고생하셨습니다~

Kminss committed:
#14 [fix-auction] 입찰하기 redis crud 수정

- Redisson 자료구조에 값을 저장할 때, 클래스 타입 정보를 같이 저장함.
 값만 저장하기 위해서는 codec에 타입 정보클래스를 초기화 하고 해당 코덱을
 세팅해줘야함
 입찰관련 redisson repository class 를 만들어 사용하기로 변경

 추후 리팩토링 여지가 많음

 - 요청한 입찰가에 대해 포인트 검증 및 차감 추가

Kminss committed:
#14 [fix-auction] 입찰하기 포인트 차감 추가

 - 요청한 입찰가에 대해 포인트 검증 및 차감 추가

Kminss committed:
#14 [bugfix-auction] 입찰하기 포인트 검증 버그 수정

Kminss committed:
#14 [fix-auction] 입찰하기 기능 - Redisson Lock 설정 수정

- Redis Lock 보다 Transaction timeout이 늦게 풀려 동시성이 깨지는 문제 수정
  => 락의 leaseTime을 TransactionTimeout보다 늦게 풀리게 time 설정

Kminss committed:
#14 [fix-auction] 입찰하기 기능 - 추가 예외처리

- lease (락을 잡고있을 수 있는시간) 이 지난 상태에서 transaction 처리가 끝나고 난 후,
  unlock 을 시도할 때, 생기는 예외 처리
  => 3e03bc4d 에서 lease 시간이 Transaction 보다 늦게 끝나게끔 설정을 해두어 가능

Kminss committed:
#14 [refactor-auction] 입찰하기 기능 - 리팩토링

Kminss committed:
#14 [init-auction] 입찰하기 기능 - Transaction timeout 설정

- RedissonLock 사용하는 Transaction 의 time out 을 설정

kimhyeyun reviewed (approved):
고생하셨습니당~

Kminss committed:
#14 [test-auction] 입찰하기 기능 - 테스트 수정

- 서비스 코드 리팩토링으로 테스트 코드 수정
